core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -195,45 +197,56 @@ private[spark] class ProcfsMetricsGetter(procfsDir: String = "/proc/") extends L
         if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
           allMetrics.copy(
             jvmVmemTotal = allMetrics.jvmVmemTotal + vmem,
-            jvmRSSTotal = allMetrics.jvmRSSTotal + (rssMem)
+            jvmRSSTotal = allMetrics.jvmRSSTotal + (rssMem),
+            timeStamp = System.currentTimeMillis
           )
         }
         else if (procInfoSplit(1).toLowerCase(Locale.US).contains("python")) {
           allMetrics.copy(
             pythonVmemTotal = allMetrics.pythonVmemTotal + vmem,
-            pythonRSSTotal = allMetrics.pythonRSSTotal + (rssMem)
+            pythonRSSTotal = allMetrics.pythonRSSTotal + (rssMem),
+            timeStamp = System.currentTimeMillis
           )
         }
         else {
           allMetrics.copy(
             otherVmemTotal = allMetrics.otherVmemTotal + vmem,
-            otherRSSTotal = allMetrics.otherRSSTotal + (rssMem)
+            otherRSSTotal = allMetrics.otherRSSTotal + (rssMem),
+            timeStamp = System.currentTimeMillis
           )
         }
       }
     } catch {
       case f: IOException =>
         logWarning("There was a problem with reading" +
           " the stat file of the process. ", f)
-        ProcfsMetrics(0, 0, 0, 0, 0, 0)
+        ProcfsMetrics(0, 0, 0, 0, 0, 0, System.currentTimeMillis)
     }
   }
 
   private[spark] def computeAllMetrics(): ProcfsMetrics = {
     if (!isAvailable) {
-      return ProcfsMetrics(0, 0, 0, 0, 0, 0)
+      return ProcfsMetrics(0, 0, 0, 0, 0, 0, System.currentTimeMillis)
     }
-    val pids = computeProcessTree
-    var allMetrics = ProcfsMetrics(0, 0, 0, 0, 0, 0)
-    for (p <- pids) {
-      allMetrics = addProcfsMetricsFromOneProcess(allMetrics, p)
-      // if we had an error getting any of the metrics, we don't want to report partial metrics, as
-      // that would be misleading.
-      if (!isAvailable) {
-        return ProcfsMetrics(0, 0, 0, 0, 0, 0)
+    val lastMetricComputation = System.currentTimeMillis() - cachedAllMetric.timeStamp
+    // Check whether we have computed the metrics in the past 1s
+    // ToDo: Should we make this configurable?
+    if(lastMetricComputation > Math.min(1000, HEARTBEAT_INTERVAL_MS)) {
```

Comment:
Also, maybe ensure that no prior computation it currently going on, just to ensure that if a user sets the configurable value too low, this does not get into a continuous loop of computing metrics

Reply:
I want to do some testing before adding a synchronized method. Metrics doesn't seems to be that critical and a benign race that removing it will cause bad performance isn't a good idea. As a last solution I think defining this method as synchronized is ok.